src/common/transformations/src/transformations/faster_hash.cpp

```diff
@@ -0,0 +1,422 @@
+// Copyright (C) 2018-2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "ngraph/ops.hpp"
+#include "ngraph/opsets/opset1.hpp"
+#include "openvino/op/util/framework_node.hpp"
+#include "transformations/hash.hpp"
+
+using namespace ngraph;
+
+namespace {  // helpers
+
+template <typename T>
+static uint64_t hash_combine(uint64_t seed, const T& a) {
+    // Hash combine formula from boost
+    return seed ^ (std::hash<T>()(a) + 0x9e3779b9 + (seed << 6) + (seed >> 2));
+}
+
+template <typename Container>
+void join_hash(uint64_t& hash, const Container& c) {
+    for (const auto& v : c) {
+        hash = hash_combine(hash, v);
+    }
+}
+
+size_t hash_combine(const void* v, int64_t size) {
+    constexpr auto cel_size = sizeof(size_t);
+    auto seed = static_cast<size_t>(size);
+    const auto data = static_cast<const size_t*>(v);
+    const auto steps = size / cel_size;
+    const auto d_end = std::next(data, steps);
+    // Slower version: this is > ~2 times slower than calculation of separate 4 sums
+    //    for (auto d = data; d < d_end; ++d) {
+    //        sum ^= *d + 0x9e3779b9 + (sum << 6) + (sum >> 2);
+    //    }
+
+    // ---- Faster version ---
+    // Calculate 4 separate independent hash sums within one cycle
+    constexpr auto fast_step = 4;
+    size_t sum[fast_step] = {0};
+    const auto fast_step_cnt = steps / fast_step;
+    const auto d_end_fast = std::next(data, fast_step_cnt * fast_step);
+    for (auto d = data; d < d_end_fast; d += fast_step) {
+        // This appears to be almost 2 times faster to calculate separate sums than do it with step=1
+        for (auto i = 0; i < fast_step; i++) {
+            sum[i] ^= d[i] + 0x9e3779b9 + (sum[i] << 6) + (sum[i] >> 2);
+        }
+    }
+    for (auto i = 0; i < steps % fast_step; i++) {
+        seed = hash_combine(seed, d_end_fast[i]);
+    }
+    for (auto i = 0; i < fast_step; i++) {
+        seed = hash_combine(seed, sum[i]);
+    }
+    // ---- End Faster version ----
+    size_t last_bytes{0};
+    std::memcpy(&last_bytes, d_end, size % cel_size);
+    seed ^= last_bytes + 0x9e3779b9 + (seed << 6) + (seed >> 2);
```

Comment:
The same code as in hash_combine

Reply:
The same, using `hash_combine` will add std::hash(last_bytes) which will bring overhead for weights hash calculation